metaflow/plugins/kfp/kfp.py

```diff
@@ -1,316 +1,395 @@
+            ]
+
+            # If the start step gets retried, we must be careful not to
+            # regenerate multiple parameters tasks. Hence we check first if
+            # _parameters exists already.
+            task_id_params_path = "{kfp_run_id}/_parameters/{task_id_params}".format(
+                kfp_run_id=kfp_run_id, task_id_params=task_id_params
+            )
+            exists = entrypoint + ["dump", "--max-value-size=0", task_id_params_path]
+            cmd = "if ! %s >/dev/null 2>/dev/null; then %s && %s; fi" % (
+                " ".join(exists),
+                export_params,
+                " ".join(params),
+            )
+            cmds.append(cmd)
+            paths = task_id_params_path
+
+        if node.type == "join" and self.graph[node.split_parents[-1]].type == "foreach":
+            # TODO: get from Dynamo or a kv store
+            parent_tasks_file = "".join(random.choice(string.ascii_lowercase) for _ in range(10))
+            # export_parent_tasks = \
+            #     'python -m ' \
+            #     'metaflow.plugins.aws.step_functions.set_batch_environment ' \
+            #     'parent_tasks %s && . `pwd`/%s' \
+            #         % (parent_tasks_file, parent_tasks_file)
+            # cmds.append(export_parent_tasks)
+            cmds.append("echo %s" % parent_tasks_file)
+
+        top_level = [
+            "--quiet",
+            "--metadata=%s" % self.metadata.TYPE,
+            "--environment=%s" % self.environment.TYPE,
+            "--datastore=%s" % self.datastore.TYPE,
+            "--datastore-root=%s" % self.datastore.datastore_root,
+            "--event-logger=%s" % self.event_logger.logger_type,
+            "--monitor=%s" % self.monitor.monitor_type,
+            "--no-pylint",
+            "--with=kfp_internal",
+        ]
+
+        step = [
+            "step",
+            node.name,
+            "--run-id %s" % kfp_run_id,
+            "--task-id %s" % task_id,
+            # Since retries are handled by KFP Argo, we can rely on
+            # {{retries}} as the job counter.
+            # '--retry-count {{retries}}',  # TODO: test verify, should it be -1?
+            "--max-user-code-retries %d" % user_code_retries,
+            "--input-paths %s" % paths,
+        ]
+
+        if any(self.graph[n].type == "foreach" for n in node.in_funcs):
+            # We set the `METAFLOW_SPLIT_INDEX` through JSONPath-foo
+            # to pass the state from the parent DynamoDb state for for-each.
+            step.append("--split-index $METAFLOW_SPLIT_INDEX")  # TODO: get from KFP param
+        if self.namespace:
+            step.append("--namespace %s" % self.namespace)
+        cmds.append(" ".join(entrypoint + top_level + step))
+        return " && ".join(cmds)
+
+    def create_kfp_pipeline_from_flow_graph(self):
+        import kfp
+        from kfp import dsl
+
+        base_image = "ssreejith3/mf_on_kfp:python-curl-git"  # TODO: environment variable
```

Comment:
Do we have a plan to publish a Docker image ourselves from the ZG Github/Docker Hub org?

Reply:
Created internal AIP-1980